In [1]:
import $ivy.`org.apache.spark::spark-core:4.1.1`
import $ivy.`org.apache.spark::spark-sql:4.1.1`

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

val spark = SparkSession.builder()
  .appName("Clase17_Sesion1_DataFrames")
  .master("local[*]")
  .getOrCreate()

import spark.implicits._
spark.sparkContext.setLogLevel("ERROR")

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
26/04/28 09:36:08 INFO SparkContext: Running Spark version 4.1.1
26/04/28 09:36:08 INFO SparkContext: OS info Windows 11, 10.0, amd64
26/04/28 09:36:08 INFO SparkContext: Java version 17.0.12+8-LTS-286
26/04/28 09:36:09 INFO ResourceUtils: ==============================================================
26/04/28 09:36:09 INFO ResourceUtils: No custom resources configured for spark.driver.
26/04/28 09:36:09 INFO ResourceUtils: ==============================================================
26/04/28 09:36:09 INFO SparkContext: Submitted application: Clase17_Sesion1_DataFrames
26/04/28 09:36:09 INFO SecurityManager: Changing view acls to: Imp_06 - Ma26/04/28 09:36:09 INFO SecurityManager: Changing view acls to: Imp_06 - Ma26/04/28 09:36:09 INFO SecurityManager: Changing view acls to: Imp_06 - Ma26/04/28 09:36:09 INFO SecurityManager: Changing view acls to: Imp_06 - Ma26/04/28 09:36:09 INFO SecurityManager: Chang

2026-04-28T07:36:09.561661500Z scala-kernel-interpreter-1 ERROR An exception occurred processing Appender console org.apache.logging.log4j.core.appender.AppenderLoggingException: java.lang.AssertionError: assertion failed
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:164)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:133)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:124)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:88)
	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:714)
	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:672)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:648)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:584)
	at org.apache.logging.log4j.

import $ivy.$
import $ivy.$
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
spark: SparkSession = org.apache.spark.sql.classic.SparkSession@4f6e12b4
import spark.implicits._

In [2]:
val dfVentas = Seq(
(1,"Laptop","Tecnología",1200.0,2,"Madrid"),
(2,"Teclado","Tecnología",45.0,5,"Valencia"),
(3,"Monitor","Tecnología",350.0,1,"Madrid"),
(4,"Silla","Oficina",120.0,4,"Barcelona"),
(5,"Mesa","Oficina",250.0,2,"Sevilla"),
(6,"Laptop","Tecnología",1200.0,1,"Madrid"),
(7,"Tablet","Tecnología",600.0,2,"Bilbao"),
(8,"Ratón","Tecnología",25.0,10,"Madrid"),
(9,"Impresora","Tecnología",200.0,1,"Sevilla"),
(10,"Silla","Oficina",120.0,3,"Madrid"),
(11,"Mesa","Oficina",250.0,1,"Barcelona"),
(12,"Laptop","Tecnología",1200.0,1,"Valencia"),
(13,"Monitor","Tecnología",350.0,2,"Madrid"),
(14,"Tablet","Tecnología",600.0,1,"Sevilla"),
(15,"Ratón","Tecnología",25.0,8,"Bilbao"),
(16,"Teclado","Tecnología",45.0,6,"Madrid"),
(17,"Silla","Oficina",120.0,2,"Valencia"),
(18,"Mesa","Oficina",250.0,3,"Madrid"),
(19,"Laptop","Tecnología",1200.0,1,"Barcelona"),
(20,"Monitor","Tecnología",350.0,1,"Sevilla"),
(21,"Tablet","Tecnología",600.0,2,"Madrid"),
(22,"Ratón","Tecnología",25.0,12,"Madrid"),
(23,"Teclado","Tecnología",45.0,4,"Bilbao"),
(24,"Silla","Oficina",120.0,1,"Sevilla"),
(25,"Mesa","Oficina",250.0,2,"Madrid"),
(26,"Laptop","Tecnología",1200.0,3,"Madrid"),
(27,"Monitor","Tecnología",350.0,2,"Barcelona"),
(28,"Tablet","Tecnología",600.0,1,"Valencia"),
(29,"Ratón","Tecnología",25.0,15,"Madrid"),
(30,"Teclado","Tecnología",45.0,7,"Sevilla"),
(31,"Silla","Oficina",120.0,5,"Madrid"),
(32,"Mesa","Oficina",250.0,1,"Bilbao"),
(33,"Laptop","Tecnología",1200.0,2,"Sevilla"),
(34,"Monitor","Tecnología",350.0,3,"Madrid"),
(35,"Tablet","Tecnología",600.0,2,"Barcelona"),
(36,"Ratón","Tecnología",25.0,9,"Valencia"),
(37,"Teclado","Tecnología",45.0,6,"Madrid"),
(38,"Silla","Oficina",120.0,2,"Bilbao"),
(39,"Mesa","Oficina",250.0,2,"Madrid"),
(40,"Laptop","Tecnología",1200.0,1,"Madrid")
).toDF("id","producto","categoria","precio","cantidad","ciudad")

dfVentas: org.apache.spark.sql.package.DataFrame = [id: int, producto: string ... 4 more fields]

In [3]:
dfVentas.show(5)

+---+--------+----------+------+--------+---------+
| id|producto| categoria|precio|cantidad|   ciudad|
+---+--------+----------+------+--------+---------+
|  1|  Laptop|Tecnología|1200.0|       2|   Madrid|
|  2| Teclado|Tecnología|  45.0|       5| Valencia|
|  3| Monitor|Tecnología| 350.0|       1|   Madrid|
|  4|   Silla|   Oficina| 120.0|       4|Barcelona|
|  5|    Mesa|   Oficina| 250.0|       2|  Sevilla|
+---+--------+----------+------+--------+---------+
only showing top 5 rows


In [4]:
dfVentas.printSchema()

root
 |-- id: integer (nullable = false)
 |-- producto: string (nullable = true)
 |-- categoria: string (nullable = true)
 |-- precio: double (nullable = false)
 |-- cantidad: integer (nullable = false)
 |-- ciudad: string (nullable = true)



In [5]:
dfVentas.describe("precio","cantidad").show()

+-------+-----------------+------------------+
|summary|           precio|          cantidad|
+-------+-----------------+------------------+
|  count|               40|                40|
|   mean|            398.0|             3.475|
| stddev|413.5831544454218|3.3203915431767985|
|    min|             25.0|                 1|
|    max|           1200.0|                15|
+-------+-----------------+------------------+



In [6]:
dfVentas.columns.foreach(println)

id
producto
categoria
precio
cantidad
ciudad


In [7]:
dfVentas.select("producto","precio").show()

+---------+------+
| producto|precio|
+---------+------+
|   Laptop|1200.0|
|  Teclado|  45.0|
|  Monitor| 350.0|
|    Silla| 120.0|
|     Mesa| 250.0|
|   Laptop|1200.0|
|   Tablet| 600.0|
|    Ratón|  25.0|
|Impresora| 200.0|
|    Silla| 120.0|
|     Mesa| 250.0|
|   Laptop|1200.0|
|  Monitor| 350.0|
|   Tablet| 600.0|
|    Ratón|  25.0|
|  Teclado|  45.0|
|    Silla| 120.0|
|     Mesa| 250.0|
|   Laptop|1200.0|
|  Monitor| 350.0|
+---------+------+
only showing top 20 rows


🧪 BLOQUE 2 — Operaciones básicas

In [8]:
dfVentas.select("producto","precio").show()

+---------+------+
| producto|precio|
+---------+------+
|   Laptop|1200.0|
|  Teclado|  45.0|
|  Monitor| 350.0|
|    Silla| 120.0|
|     Mesa| 250.0|
|   Laptop|1200.0|
|   Tablet| 600.0|
|    Ratón|  25.0|
|Impresora| 200.0|
|    Silla| 120.0|
|     Mesa| 250.0|
|   Laptop|1200.0|
|  Monitor| 350.0|
|   Tablet| 600.0|
|    Ratón|  25.0|
|  Teclado|  45.0|
|    Silla| 120.0|
|     Mesa| 250.0|
|   Laptop|1200.0|
|  Monitor| 350.0|
+---------+------+
only showing top 20 rows


In [9]:
dfVentas.filter($"precio" > 500).show()

+---+--------+----------+------+--------+---------+
| id|producto| categoria|precio|cantidad|   ciudad|
+---+--------+----------+------+--------+---------+
|  1|  Laptop|Tecnología|1200.0|       2|   Madrid|
|  6|  Laptop|Tecnología|1200.0|       1|   Madrid|
|  7|  Tablet|Tecnología| 600.0|       2|   Bilbao|
| 12|  Laptop|Tecnología|1200.0|       1| Valencia|
| 14|  Tablet|Tecnología| 600.0|       1|  Sevilla|
| 19|  Laptop|Tecnología|1200.0|       1|Barcelona|
| 21|  Tablet|Tecnología| 600.0|       2|   Madrid|
| 26|  Laptop|Tecnología|1200.0|       3|   Madrid|
| 28|  Tablet|Tecnología| 600.0|       1| Valencia|
| 33|  Laptop|Tecnología|1200.0|       2|  Sevilla|
| 35|  Tablet|Tecnología| 600.0|       2|Barcelona|
| 40|  Laptop|Tecnología|1200.0|       1|   Madrid|
+---+--------+----------+------+--------+---------+



In [10]:
val dfTotal = dfVentas.withColumn("total", $"precio" * $"cantidad")

dfTotal: org.apache.spark.sql.package.DataFrame = [id: int, producto: string ... 5 more fields]

In [11]:
dfTotal.orderBy($"total".desc).show()

+---+--------+----------+------+--------+---------+------+
| id|producto| categoria|precio|cantidad|   ciudad| total|
+---+--------+----------+------+--------+---------+------+
| 26|  Laptop|Tecnología|1200.0|       3|   Madrid|3600.0|
| 33|  Laptop|Tecnología|1200.0|       2|  Sevilla|2400.0|
|  1|  Laptop|Tecnología|1200.0|       2|   Madrid|2400.0|
| 19|  Laptop|Tecnología|1200.0|       1|Barcelona|1200.0|
| 35|  Tablet|Tecnología| 600.0|       2|Barcelona|1200.0|
|  6|  Laptop|Tecnología|1200.0|       1|   Madrid|1200.0|
| 12|  Laptop|Tecnología|1200.0|       1| Valencia|1200.0|
|  7|  Tablet|Tecnología| 600.0|       2|   Bilbao|1200.0|
| 21|  Tablet|Tecnología| 600.0|       2|   Madrid|1200.0|
| 40|  Laptop|Tecnología|1200.0|       1|   Madrid|1200.0|
| 34| Monitor|Tecnología| 350.0|       3|   Madrid|1050.0|
| 18|    Mesa|   Oficina| 250.0|       3|   Madrid| 750.0|
| 13| Monitor|Tecnología| 350.0|       2|   Madrid| 700.0|
| 27| Monitor|Tecnología| 350.0|       2|Barcelona| 700.

🧪 Ejercicios basados en preguntas de negocio.

In [14]:
dfVentas.filter($"ciudad" === "Madrid").show()

+---+--------+----------+------+--------+------+
| id|producto| categoria|precio|cantidad|ciudad|
+---+--------+----------+------+--------+------+
|  1|  Laptop|Tecnología|1200.0|       2|Madrid|
|  3| Monitor|Tecnología| 350.0|       1|Madrid|
|  6|  Laptop|Tecnología|1200.0|       1|Madrid|
|  8|   Ratón|Tecnología|  25.0|      10|Madrid|
| 10|   Silla|   Oficina| 120.0|       3|Madrid|
| 13| Monitor|Tecnología| 350.0|       2|Madrid|
| 16| Teclado|Tecnología|  45.0|       6|Madrid|
| 18|    Mesa|   Oficina| 250.0|       3|Madrid|
| 21|  Tablet|Tecnología| 600.0|       2|Madrid|
| 22|   Ratón|Tecnología|  25.0|      12|Madrid|
| 25|    Mesa|   Oficina| 250.0|       2|Madrid|
| 26|  Laptop|Tecnología|1200.0|       3|Madrid|
| 29|   Ratón|Tecnología|  25.0|      15|Madrid|
| 31|   Silla|   Oficina| 120.0|       5|Madrid|
| 34| Monitor|Tecnología| 350.0|       3|Madrid|
| 37| Teclado|Tecnología|  45.0|       6|Madrid|
| 39|    Mesa|   Oficina| 250.0|       2|Madrid|
| 40|  Laptop|Tecnol

In [15]:
dfVentas.withColumn("precio_iva", $"precio" * 1.21).show()

+---+---------+----------+------+--------+---------+------------------+
| id| producto| categoria|precio|cantidad|   ciudad|        precio_iva|
+---+---------+----------+------+--------+---------+------------------+
|  1|   Laptop|Tecnología|1200.0|       2|   Madrid|            1452.0|
|  2|  Teclado|Tecnología|  45.0|       5| Valencia|54.449999999999996|
|  3|  Monitor|Tecnología| 350.0|       1|   Madrid|             423.5|
|  4|    Silla|   Oficina| 120.0|       4|Barcelona|             145.2|
|  5|     Mesa|   Oficina| 250.0|       2|  Sevilla|             302.5|
|  6|   Laptop|Tecnología|1200.0|       1|   Madrid|            1452.0|
|  7|   Tablet|Tecnología| 600.0|       2|   Bilbao|             726.0|
|  8|    Ratón|Tecnología|  25.0|      10|   Madrid|             30.25|
|  9|Impresora|Tecnología| 200.0|       1|  Sevilla|             242.0|
| 10|    Silla|   Oficina| 120.0|       3|   Madrid|             145.2|
| 11|     Mesa|   Oficina| 250.0|       1|Barcelona|            

In [16]:
dfVentas.drop("categoria").show()

+---+---------+------+--------+---------+
| id| producto|precio|cantidad|   ciudad|
+---+---------+------+--------+---------+
|  1|   Laptop|1200.0|       2|   Madrid|
|  2|  Teclado|  45.0|       5| Valencia|
|  3|  Monitor| 350.0|       1|   Madrid|
|  4|    Silla| 120.0|       4|Barcelona|
|  5|     Mesa| 250.0|       2|  Sevilla|
|  6|   Laptop|1200.0|       1|   Madrid|
|  7|   Tablet| 600.0|       2|   Bilbao|
|  8|    Ratón|  25.0|      10|   Madrid|
|  9|Impresora| 200.0|       1|  Sevilla|
| 10|    Silla| 120.0|       3|   Madrid|
| 11|     Mesa| 250.0|       1|Barcelona|
| 12|   Laptop|1200.0|       1| Valencia|
| 13|  Monitor| 350.0|       2|   Madrid|
| 14|   Tablet| 600.0|       1|  Sevilla|
| 15|    Ratón|  25.0|       8|   Bilbao|
| 16|  Teclado|  45.0|       6|   Madrid|
| 17|    Silla| 120.0|       2| Valencia|
| 18|     Mesa| 250.0|       3|   Madrid|
| 19|   Laptop|1200.0|       1|Barcelona|
| 20|  Monitor| 350.0|       1|  Sevilla|
+---+---------+------+--------+---

In [17]:
dfVentas.withColumnRenamed("precio","precio_unitario").show()

+---+---------+----------+---------------+--------+---------+
| id| producto| categoria|precio_unitario|cantidad|   ciudad|
+---+---------+----------+---------------+--------+---------+
|  1|   Laptop|Tecnología|         1200.0|       2|   Madrid|
|  2|  Teclado|Tecnología|           45.0|       5| Valencia|
|  3|  Monitor|Tecnología|          350.0|       1|   Madrid|
|  4|    Silla|   Oficina|          120.0|       4|Barcelona|
|  5|     Mesa|   Oficina|          250.0|       2|  Sevilla|
|  6|   Laptop|Tecnología|         1200.0|       1|   Madrid|
|  7|   Tablet|Tecnología|          600.0|       2|   Bilbao|
|  8|    Ratón|Tecnología|           25.0|      10|   Madrid|
|  9|Impresora|Tecnología|          200.0|       1|  Sevilla|
| 10|    Silla|   Oficina|          120.0|       3|   Madrid|
| 11|     Mesa|   Oficina|          250.0|       1|Barcelona|
| 12|   Laptop|Tecnología|         1200.0|       1| Valencia|
| 13|  Monitor|Tecnología|          350.0|       2|   Madrid|
| 14|   

In [18]:
dfVentas.dropDuplicates("producto").show()

+---+---------+----------+------+--------+---------+
| id| producto| categoria|precio|cantidad|   ciudad|
+---+---------+----------+------+--------+---------+
|  9|Impresora|Tecnología| 200.0|       1|  Sevilla|
|  1|   Laptop|Tecnología|1200.0|       2|   Madrid|
|  5|     Mesa|   Oficina| 250.0|       2|  Sevilla|
|  3|  Monitor|Tecnología| 350.0|       1|   Madrid|
|  8|    Ratón|Tecnología|  25.0|      10|   Madrid|
|  4|    Silla|   Oficina| 120.0|       4|Barcelona|
|  7|   Tablet|Tecnología| 600.0|       2|   Bilbao|
|  2|  Teclado|Tecnología|  45.0|       5| Valencia|
+---+---------+----------+------+--------+---------+



In [19]:
dfVentas.count()

res19: Long = 40L

In [20]:
dfVentas.columns

res20: Array[String] = Array(
  "id",
  "producto",
  "categoria",
  "precio",
  "cantidad",
  "ciudad"
)

In [21]:
dfVentas.dtypes.foreach(println)

(id,IntegerType)
(producto,StringType)
(categoria,StringType)
(precio,DoubleType)
(cantidad,IntegerType)
(ciudad,StringType)


In [22]:
dfVentas.describe("precio").show()

+-------+-----------------+
|summary|           precio|
+-------+-----------------+
|  count|               40|
|   mean|            398.0|
| stddev|413.5831544454218|
|    min|             25.0|
|    max|           1200.0|
+-------+-----------------+



🧪 PIPELINES

 Enunciado: Queremos analizar las ventas de mayor valor económico. Construye un pipeline que:

1. Filtre solo los productos cuyo precio sea mayor a 100
2. Cree una nueva columna llamada total (precio × cantidad)
3. Seleccione únicamente las columnas:
    * producto
    * total
    * ciudad
4. Ordene los resultados de mayor a menor total

In [23]:
val ventasMayorValor = dfVentas
  .filter($"precio" > 100) // filtra los productos con precio mayor a 100.
  .withColumn("total", $"precio" * $"cantidad") // Se crea la nueva columna total con el calculo de precio * total.
  .select("producto", "total", "ciudad") // Se selecciona las columnas producto, total y ciudad.
  .orderBy($"total".desc) // Se ordena por el total de forma descendente.

ventasMayorValor.show()

+--------+------+---------+
|producto| total|   ciudad|
+--------+------+---------+
|  Laptop|3600.0|   Madrid|
|  Laptop|2400.0|  Sevilla|
|  Laptop|2400.0|   Madrid|
|  Tablet|1200.0|Barcelona|
|  Tablet|1200.0|   Madrid|
|  Laptop|1200.0|   Madrid|
|  Laptop|1200.0|   Madrid|
|  Laptop|1200.0| Valencia|
|  Laptop|1200.0|Barcelona|
|  Tablet|1200.0|   Bilbao|
| Monitor|1050.0|   Madrid|
|    Mesa| 750.0|   Madrid|
| Monitor| 700.0|Barcelona|
| Monitor| 700.0|   Madrid|
|  Tablet| 600.0| Valencia|
|   Silla| 600.0|   Madrid|
|  Tablet| 600.0|  Sevilla|
|    Mesa| 500.0|   Madrid|
|    Mesa| 500.0|   Madrid|
|    Mesa| 500.0|  Sevilla|
+--------+------+---------+
only showing top 20 rows


ventasMayorValor: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [producto: string, total: double ... 1 more field]

👉 Enunciado: Queremos obtener una lista de productos únicos junto con su precio más representativo. Construye un pipeline que:

1. Elimine los duplicados basándose en la columna producto
2. Seleccione únicamente:
    * producto
    * precio
3. Ordene los resultados de mayor a menor precio

In [24]:
val productosUnicos = dfVentas
  .dropDuplicates("producto") // Se eliminan los duplicados por producto
  .select("producto", "precio") // se selecciona solo los campos producto y precio
  .orderBy($"precio".desc) // se ordena por el precio de forma descedente

productosUnicos.show()

+---------+------+
| producto|precio|
+---------+------+
|   Laptop|1200.0|
|   Tablet| 600.0|
|  Monitor| 350.0|
|     Mesa| 250.0|
|Impresora| 200.0|
|    Silla| 120.0|
|  Teclado|  45.0|
|    Ratón|  25.0|
+---------+------+



productosUnicos: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [producto: string, precio: double]

🏢 Caso de Estudio — MobilityData Analytics

In [25]:
val viajes = Seq(
(1,"2026-05-01","Madrid",12.5,10.0),
(2,"2026-05-01","Barcelona",8.0,7.0),
(3,"2026-05-02","Madrid",15.0,12.0),
(4,"2026-05-02","Valencia",6.0,6.0),
(5,"2026-05-03","Sevilla",20.0,15.0),
(6,"2026-05-03","Madrid",5.0,5.0),
(7,"2026-05-04","Bilbao",18.0,13.0),
(8,"2026-05-04","Madrid",22.0,16.0),
(9,"2026-05-05","Barcelona",10.0,9.0),
(10,"2026-05-05","Madrid",14.0,11.0),
(11,"2026-05-06","Valencia",7.0,6.5),
(12,"2026-05-06","Madrid",16.0,13.0),
(13,"2026-05-07","Sevilla",25.0,18.0),
(14,"2026-05-07","Madrid",9.0,8.0),
(15,"2026-05-08","Barcelona",11.0,9.5),
(16,"2026-05-08","Bilbao",13.0,10.0),
(17,"2026-05-09","Madrid",17.0,14.0),
(18,"2026-05-09","Valencia",8.0,7.5),
(19,"2026-05-10","Sevilla",19.0,14.0),
(20,"2026-05-10","Madrid",21.0,17.0),
(21,"2026-05-11","Barcelona",9.0,8.0),
(22,"2026-05-11","Madrid",18.0,15.0),
(23,"2026-05-12","Valencia",6.5,6.0),
(24,"2026-05-12","Madrid",20.0,16.0),
(25,"2026-05-13","Bilbao",14.0,11.0),
(26,"2026-05-13","Madrid",23.0,18.0),
(27,"2026-05-14","Sevilla",18.0,13.0),
(28,"2026-05-14","Madrid",7.0,6.0),
(29,"2026-05-15","Barcelona",12.0,10.0),
(30,"2026-05-15","Madrid",19.0,15.0)
).toDF("id","fecha","ciudad","distancia_km","precio")

viajes: org.apache.spark.sql.package.DataFrame = [id: int, fecha: string ... 3 more fields]

A partir del dataset de viajes, construye un pipeline de datos que:

1. Calcule el ingreso de cada viaje: (distancia × precio)
2. Filtre únicamente los viajes con ingresos altos (> 150)
3. Seleccione solo las columnas relevantes:
    * ciudad
    * ingreso_viaje
4. Ordene los resultados de mayor a menor ingreso

In [ ]:
val pipelineViaje = viajes
  .withColumn("ingreso_viaje", $"distancia_km" * $"precio") 
  .filter($"ingreso_viaje" > 150) 
  .select("ciudad", "ingreso_viaje") 
  .orderBy($"ingreso_viaje".desc) // Se ordena por el total de forma descendente.

pipelineViaje.show()

+-------+-------------+
| ciudad|ingreso_viaje|
+-------+-------------+
|Sevilla|        450.0|
| Madrid|        414.0|
| Madrid|        357.0|
| Madrid|        352.0|
| Madrid|        320.0|
|Sevilla|        300.0|
| Madrid|        285.0|
| Madrid|        270.0|
|Sevilla|        266.0|
| Madrid|        238.0|
| Bilbao|        234.0|
|Sevilla|        234.0|
| Madrid|        208.0|
| Madrid|        180.0|
| Madrid|        154.0|
| Bilbao|        154.0|
+-------+-------------+



pipelineViaje: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, ingreso_viaje: double]

In [33]:
println("1. ¿Cuántos viajes hay?")
viajes.count()

1. ¿Cuántos viajes hay?


res33_1: Long = 30L

In [34]:
println("*"*25)
println("2. ¿Qué columnas existen?")
viajes.columns

*************************
2. ¿Qué columnas existen?


res34_2: Array[String] = Array(
  "id",
  "fecha",
  "ciudad",
  "distancia_km",
  "precio"
)

In [35]:
println("*"*25)
println("3. ¿Qué tipos tienen?")
viajes.dtypes.foreach(println)

*************************
3. ¿Qué tipos tienen?
(id,IntegerType)
(fecha,StringType)
(ciudad,StringType)
(distancia_km,DoubleType)
(precio,DoubleType)


🔹 Transformación
1. ¿Cuántos viajes cumplen el filtro?
2. ¿Cuál tiene mayor ingreso?
3. Muestra solo viajes de Madrid

In [36]:
println("Cuantos viajes cumplen el filtro")
pipelineViaje.count()

Cuantos viajes cumplen el filtro


res36_1: Long = 16L

In [37]:
println("cual tiene mayor ingreso")
pipelineViaje.show(1)

cual tiene mayor ingreso
+-------+-------------+
| ciudad|ingreso_viaje|
+-------+-------------+
|Sevilla|        450.0|
+-------+-------------+
only showing top 1 row


In [38]:
println("mostrar solo viajes de madrid")
pipelineViaje.filter($"ciudad" === "Madrid").show()

mostrar solo viajes de madrid
+------+-------------+
|ciudad|ingreso_viaje|
+------+-------------+
|Madrid|        414.0|
|Madrid|        357.0|
|Madrid|        352.0|
|Madrid|        320.0|
|Madrid|        285.0|
|Madrid|        270.0|
|Madrid|        238.0|
|Madrid|        208.0|
|Madrid|        180.0|
|Madrid|        154.0|
+------+-------------+



🔹 Manipulación
1. Elimina columna distancia
2. Renombra precio
3. Quita duplicados por ciudad

In [39]:
viajes.drop("distancia_km").show()

+---+----------+---------+------+
| id|     fecha|   ciudad|precio|
+---+----------+---------+------+
|  1|2026-05-01|   Madrid|  10.0|
|  2|2026-05-01|Barcelona|   7.0|
|  3|2026-05-02|   Madrid|  12.0|
|  4|2026-05-02| Valencia|   6.0|
|  5|2026-05-03|  Sevilla|  15.0|
|  6|2026-05-03|   Madrid|   5.0|
|  7|2026-05-04|   Bilbao|  13.0|
|  8|2026-05-04|   Madrid|  16.0|
|  9|2026-05-05|Barcelona|   9.0|
| 10|2026-05-05|   Madrid|  11.0|
| 11|2026-05-06| Valencia|   6.5|
| 12|2026-05-06|   Madrid|  13.0|
| 13|2026-05-07|  Sevilla|  18.0|
| 14|2026-05-07|   Madrid|   8.0|
| 15|2026-05-08|Barcelona|   9.5|
| 16|2026-05-08|   Bilbao|  10.0|
| 17|2026-05-09|   Madrid|  14.0|
| 18|2026-05-09| Valencia|   7.5|
| 19|2026-05-10|  Sevilla|  14.0|
| 20|2026-05-10|   Madrid|  17.0|
+---+----------+---------+------+
only showing top 20 rows


In [40]:
viajes.withColumnRenamed("precio","precio_unitario").show()

+---+----------+---------+------------+---------------+
| id|     fecha|   ciudad|distancia_km|precio_unitario|
+---+----------+---------+------------+---------------+
|  1|2026-05-01|   Madrid|        12.5|           10.0|
|  2|2026-05-01|Barcelona|         8.0|            7.0|
|  3|2026-05-02|   Madrid|        15.0|           12.0|
|  4|2026-05-02| Valencia|         6.0|            6.0|
|  5|2026-05-03|  Sevilla|        20.0|           15.0|
|  6|2026-05-03|   Madrid|         5.0|            5.0|
|  7|2026-05-04|   Bilbao|        18.0|           13.0|
|  8|2026-05-04|   Madrid|        22.0|           16.0|
|  9|2026-05-05|Barcelona|        10.0|            9.0|
| 10|2026-05-05|   Madrid|        14.0|           11.0|
| 11|2026-05-06| Valencia|         7.0|            6.5|
| 12|2026-05-06|   Madrid|        16.0|           13.0|
| 13|2026-05-07|  Sevilla|        25.0|           18.0|
| 14|2026-05-07|   Madrid|         9.0|            8.0|
| 15|2026-05-08|Barcelona|        11.0|         

In [41]:
viajes.dropDuplicates("ciudad").show()

+---+----------+---------+------------+------+
| id|     fecha|   ciudad|distancia_km|precio|
+---+----------+---------+------------+------+
|  2|2026-05-01|Barcelona|         8.0|   7.0|
|  7|2026-05-04|   Bilbao|        18.0|  13.0|
|  1|2026-05-01|   Madrid|        12.5|  10.0|
|  5|2026-05-03|  Sevilla|        20.0|  15.0|
|  4|2026-05-02| Valencia|         6.0|   6.0|
+---+----------+---------+------------+------+



🔹 Análisis
1. Ordena por precio
2. Cambia filtro a >100
3. ¿Qué ciudad aparece más?

In [ ]:
val pipelineViaje2 = viajes
  .withColumn("ingreso_viaje", $"distancia_km" * $"precio") 
  .filter($"ingreso_viaje" > 100) 
  .select("ciudad", "ingreso_viaje") 
  .orderBy($"ingreso_viaje".desc) // Se ordena por el total de forma descendente.

pipelineViaje2.show() //Solo aparecen 4

+---------+-------------+
|   ciudad|ingreso_viaje|
+---------+-------------+
|  Sevilla|        450.0|
|   Madrid|        414.0|
|   Madrid|        357.0|
|   Madrid|        352.0|
|   Madrid|        320.0|
|  Sevilla|        300.0|
|   Madrid|        285.0|
|   Madrid|        270.0|
|  Sevilla|        266.0|
|   Madrid|        238.0|
|   Bilbao|        234.0|
|  Sevilla|        234.0|
|   Madrid|        208.0|
|   Madrid|        180.0|
|   Madrid|        154.0|
|   Bilbao|        154.0|
|   Bilbao|        130.0|
|   Madrid|        125.0|
|Barcelona|        120.0|
|Barcelona|        104.5|
+---------+-------------+



pipelineViaje2: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, ingreso_viaje: double]

In [46]:
pipelineViaje.write
  .mode("overwrite")
  .option("header", "true")
  .csv("C:/Curso-Scala/salida/viajes_ingresos_altos")

🔹 Crear y guardar otro pipeline:
Crea un nuevo pipeline que haga lo siguiente:

1. Crear una columna llamada precio_por_km
    Fórmula: precio / distancia_km
2. Filtrar solo los viajes donde precio_por_km sea mayor que 0.8
3. Seleccionar solo las columnas:
    ciudad
    fecha
    precio_por_km
4. Ordenar los resultados de mayor a menor precio_por_km
5. Guardar el resultado en formato CSV en la ruta:
    C:/Curso-Scala/salidas/viajes_precio_por_km // Usa tu ruta


In [47]:
val resultadoViajes = viajes
  .withColumn("precio_por_km", $"precio" / $"distancia_km")
  .filter($"precio_por_km" > 0.8)
  .select("ciudad", "fecha", "precio_por_km")
  .orderBy($"precio_por_km".desc)

resultadoViajes.write
  .mode("overwrite")
  .option("header", "true")
  .csv("C:/Curso-Scala/salidas/viajes_precio_por_km")

resultadoViajes: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, fecha: string ... 1 more field]

In [48]:
resultadoViajes.show()

+---------+----------+------------------+
|   ciudad|     fecha|     precio_por_km|
+---------+----------+------------------+
| Valencia|2026-05-02|               1.0|
|   Madrid|2026-05-03|               1.0|
| Valencia|2026-05-09|            0.9375|
| Valencia|2026-05-06|0.9285714285714286|
| Valencia|2026-05-12|0.9230769230769231|
|Barcelona|2026-05-05|               0.9|
|   Madrid|2026-05-07|0.8888888888888888|
|Barcelona|2026-05-11|0.8888888888888888|
|Barcelona|2026-05-01|             0.875|
|Barcelona|2026-05-08|0.8636363636363636|
|   Madrid|2026-05-14|0.8571428571428571|
|   Madrid|2026-05-11|0.8333333333333334|
|Barcelona|2026-05-15|0.8333333333333334|
|   Madrid|2026-05-09|0.8235294117647058|
|   Madrid|2026-05-06|            0.8125|
|   Madrid|2026-05-10|0.8095238095238095|
+---------+----------+------------------+



Sesión 2 : Agregaciones, agrupaciones, ventanas, rollup, cube y pivot

In [49]:
import $ivy.`org.apache.spark::spark-core:4.1.1`
import $ivy.`org.apache.spark::spark-sql:4.1.1`

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.expressions.Window

val spark = SparkSession.builder()
  .appName("Teoria_Agregaciones_Dia15")
  .master("local[*]")
  .getOrCreate()

import spark.implicits._
spark.sparkContext.setLogLevel("ERROR")

import $ivy.$
import $ivy.$
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.expressions.Window
spark: SparkSession = org.apache.spark.sql.classic.SparkSession@4f6e12b4
import spark.implicits._

In [50]:
val dfEjemplo = Seq(
  (1, "Madrid", "Tecnología", "Laptop", 2400.0),
  (2, "Madrid", "Tecnología", "Monitor", 700.0),
  (3, "Valencia", "Oficina", "Silla", 480.0),
  (4, "Valencia", "Tecnología", "Teclado", 270.0)
).toDF("id", "ciudad", "categoria", "producto", "total")

dfEjemplo.show()

+---+--------+----------+--------+------+
| id|  ciudad| categoria|producto| total|
+---+--------+----------+--------+------+
|  1|  Madrid|Tecnología|  Laptop|2400.0|
|  2|  Madrid|Tecnología| Monitor| 700.0|
|  3|Valencia|   Oficina|   Silla| 480.0|
|  4|Valencia|Tecnología| Teclado| 270.0|
+---+--------+----------+--------+------+



dfEjemplo: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 3 more fields]

In [51]:
dfEjemplo
  .groupBy("ciudad")
  .agg(sum("total").as("total_ciudad"))
  .show()

+--------+------------+
|  ciudad|total_ciudad|
+--------+------------+
|  Madrid|      3100.0|
|Valencia|       750.0|
+--------+------------+



In [52]:
val dfGroupBy = Seq(
  (1, "Madrid", "Laptop"),
  (2, "Madrid", "Monitor"),
  (3, "Valencia", "Silla"),
  (4, "Madrid", "Teclado"),
  (5, "Valencia", "Mesa")
).toDF("id", "ciudad", "producto")

dfGroupBy.show()

+---+--------+--------+
| id|  ciudad|producto|
+---+--------+--------+
|  1|  Madrid|  Laptop|
|  2|  Madrid| Monitor|
|  3|Valencia|   Silla|
|  4|  Madrid| Teclado|
|  5|Valencia|    Mesa|
+---+--------+--------+



dfGroupBy: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 1 more field]

In [53]:
dfGroupBy
  .groupBy("ciudad")
  .count()
  .show()

+--------+-----+
|  ciudad|count|
+--------+-----+
|  Madrid|    3|
|Valencia|    2|
+--------+-----+



In [54]:
import org.apache.spark.sql.functions._

import org.apache.spark.sql.functions._

In [55]:
val dfAgregaciones = Seq(
  (1, "Madrid", "Tecnología", 2400.0),
  (2, "Madrid", "Tecnología", 700.0),
  (3, "Madrid", "Oficina", 250.0),
  (4, "Valencia", "Oficina", 480.0),
  (5, "Valencia", "Tecnología", 270.0)
).toDF("id", "ciudad", "categoria", "total")

dfAgregaciones.show()

+---+--------+----------+------+
| id|  ciudad| categoria| total|
+---+--------+----------+------+
|  1|  Madrid|Tecnología|2400.0|
|  2|  Madrid|Tecnología| 700.0|
|  3|  Madrid|   Oficina| 250.0|
|  4|Valencia|   Oficina| 480.0|
|  5|Valencia|Tecnología| 270.0|
+---+--------+----------+------+



dfAgregaciones: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 2 more fields]

In [56]:
dfAgregaciones
  .groupBy("ciudad")
  .count()
  .show()

+--------+-----+
|  ciudad|count|
+--------+-----+
|  Madrid|    3|
|Valencia|    2|
+--------+-----+



In [57]:
dfAgregaciones
  .groupBy("ciudad")
  .sum("total")
  .show()

+--------+----------+
|  ciudad|sum(total)|
+--------+----------+
|  Madrid|    3350.0|
|Valencia|     750.0|
+--------+----------+



In [58]:
dfAgregaciones
  .groupBy("categoria")
  .avg("total")
  .show()

+----------+------------------+
| categoria|        avg(total)|
+----------+------------------+
|Tecnología|1123.3333333333333|
|   Oficina|             365.0|
+----------+------------------+



In [59]:
val dfAgg = Seq(
  (1, "Madrid", 2400.0),
  (2, "Madrid", 700.0),
  (3, "Madrid", 250.0),
  (4, "Valencia", 480.0),
  (5, "Valencia", 270.0)
).toDF("id", "ciudad", "total")

dfAgg.show()

+---+--------+------+
| id|  ciudad| total|
+---+--------+------+
|  1|  Madrid|2400.0|
|  2|  Madrid| 700.0|
|  3|  Madrid| 250.0|
|  4|Valencia| 480.0|
|  5|Valencia| 270.0|
+---+--------+------+



dfAgg: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 1 more field]

In [60]:
dfAgg
  .groupBy("ciudad")
  .agg(
    count("id").as("num_ventas"),
    sum("total").as("total_facturado"),
    avg("total").as("ticket_medio"),
    max("total").as("venta_maxima")
  )
  .show()

+--------+----------+---------------+------------------+------------+
|  ciudad|num_ventas|total_facturado|      ticket_medio|venta_maxima|
+--------+----------+---------------+------------------+------------+
|  Madrid|         3|         3350.0|1116.6666666666667|      2400.0|
|Valencia|         2|          750.0|             375.0|       480.0|
+--------+----------+---------------+------------------+------------+



In [61]:
val dfAlias = Seq(
  (1, "Madrid", 2400.0),
  (2, "Madrid", 700.0),
  (3, "Valencia", 480.0)
).toDF("id", "ciudad", "total")

dfAlias.show()

+---+--------+------+
| id|  ciudad| total|
+---+--------+------+
|  1|  Madrid|2400.0|
|  2|  Madrid| 700.0|
|  3|Valencia| 480.0|
+---+--------+------+



dfAlias: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 1 more field]

In [62]:
dfAlias
  .groupBy("ciudad")
  .sum("total")
  .show()

+--------+----------+
|  ciudad|sum(total)|
+--------+----------+
|  Madrid|    3100.0|
|Valencia|     480.0|
+--------+----------+



In [63]:
dfAlias
  .groupBy("ciudad")
  .agg(sum("total").as("total_facturado"))
  .show()

+--------+---------------+
|  ciudad|total_facturado|
+--------+---------------+
|  Madrid|         3100.0|
|Valencia|          480.0|
+--------+---------------+



In [64]:
import org.apache.spark.sql.expressions.Window

import org.apache.spark.sql.expressions.Window

In [65]:
val dfVentana = Seq(
  (1, "Madrid", "Laptop", 2400.0),
  (2, "Madrid", "Monitor", 700.0),
  (3, "Madrid", "Mesa", 250.0),
  (4, "Valencia", "Silla", 480.0),
  (5, "Valencia", "Teclado", 270.0)
).toDF("id", "ciudad", "producto", "total")

dfVentana.show()

+---+--------+--------+------+
| id|  ciudad|producto| total|
+---+--------+--------+------+
|  1|  Madrid|  Laptop|2400.0|
|  2|  Madrid| Monitor| 700.0|
|  3|  Madrid|    Mesa| 250.0|
|  4|Valencia|   Silla| 480.0|
|  5|Valencia| Teclado| 270.0|
+---+--------+--------+------+



dfVentana: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 2 more fields]

In [ ]:
val ventanaCiudad = Window
  .partitionBy("ciudad") //crea grupos por ciudad
  .orderBy($"total".desc)

val rankingCiudad = dfVentana
  .withColumn("ranking_ciudad", row_number().over(ventanaCiudad)) //asigna ranking dentro de cada ciudad
  .orderBy($"ciudad", $"ranking_ciudad")

rankingCiudad.show()

+---+--------+--------+------+--------------+
| id|  ciudad|producto| total|ranking_ciudad|
+---+--------+--------+------+--------------+
|  1|  Madrid|  Laptop|2400.0|             1|
|  2|  Madrid| Monitor| 700.0|             2|
|  3|  Madrid|    Mesa| 250.0|             3|
|  4|Valencia|   Silla| 480.0|             1|
|  5|Valencia| Teclado| 270.0|             2|
+---+--------+--------+------+--------------+



ventanaCiudad: org.apache.spark.sql.expressions.WindowSpec = org.apache.spark.sql.expressions.WindowSpec@3ee0c4cf
rankingCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [id: int, ciudad: string ... 3 more fields]

In [67]:
val dfRollup = Seq(
  (1, "Madrid", "Tecnología", 2400.0),
  (2, "Madrid", "Tecnología", 700.0),
  (3, "Madrid", "Oficina", 250.0),
  (4, "Valencia", "Oficina", 480.0),
  (5, "Valencia", "Tecnología", 270.0)
).toDF("id", "ciudad", "categoria", "total")

dfRollup.show()

+---+--------+----------+------+
| id|  ciudad| categoria| total|
+---+--------+----------+------+
|  1|  Madrid|Tecnología|2400.0|
|  2|  Madrid|Tecnología| 700.0|
|  3|  Madrid|   Oficina| 250.0|
|  4|Valencia|   Oficina| 480.0|
|  5|Valencia|Tecnología| 270.0|
+---+--------+----------+------+



dfRollup: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 2 more fields]

Una empresa registra sus ventas por ciudad y categoría de producto.

Se desea analizar la facturación de forma jerárquica para obtener:

1. El total facturado por cada combinación de ciudad y categoría.
2. El subtotal de facturación por ciudad, sin distinguir categorías.
3. El total general de facturación considerando todos los registros.

Para ello, utiliza la función rollup("ciudad", "categoria") junto con una agregación sobre la columna total, de modo que el resultado muestre tanto los detalles como los subtotales y el total general.

In [68]:
dfRollup
  .rollup("ciudad", "categoria")
  .agg(sum("total").as("total_facturado"))
  .orderBy($"ciudad", $"categoria")
  .show()

+--------+----------+---------------+
|  ciudad| categoria|total_facturado|
+--------+----------+---------------+
|    NULL|      NULL|         4100.0|
|  Madrid|      NULL|         3350.0|
|  Madrid|   Oficina|          250.0|
|  Madrid|Tecnología|         3100.0|
|Valencia|      NULL|          750.0|
|Valencia|   Oficina|          480.0|
|Valencia|Tecnología|          270.0|
+--------+----------+---------------+



10. cube() — combinaciones multidimensionales

In [69]:
val dfCube = Seq(
  (1, "Madrid", "Tecnología", 2400.0),
  (2, "Madrid", "Tecnología", 700.0),
  (3, "Madrid", "Oficina", 250.0),
  (4, "Valencia", "Oficina", 480.0),
  (5, "Valencia", "Tecnología", 270.0)
).toDF("id", "ciudad", "categoria", "total")

dfCube.show()

+---+--------+----------+------+
| id|  ciudad| categoria| total|
+---+--------+----------+------+
|  1|  Madrid|Tecnología|2400.0|
|  2|  Madrid|Tecnología| 700.0|
|  3|  Madrid|   Oficina| 250.0|
|  4|Valencia|   Oficina| 480.0|
|  5|Valencia|Tecnología| 270.0|
+---+--------+----------+------+



dfCube: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 2 more fields]

Ejemplo
Una empresa registra sus ventas por ciudad y categoría de producto.

La dirección desea analizar la facturación desde diferentes perspectivas, no solo por la combinación ciudad-categoría, sino también por cada dimensión de forma independiente.

Se necesita calcular:

1. El total facturado por cada combinación de ciudad y categoría.
2. El subtotal facturado por cada ciudad, sin distinguir la categoría.
3. El subtotal facturado por cada categoría, sin distinguir la ciudad.
4. El total general facturado considerando todas las ciudades y todas las categorías.

Para resolver este análisis multidimensional, utiliza la función cube("ciudad", "categoria") junto con una agregación sobre la columna total. Esta función permitirá obtener todas las combinaciones posibles de agrupación entre las columnas seleccionadas.

In [70]:
dfCube
  .cube("ciudad", "categoria")
  .agg(sum("total").as("total_facturado"))
  .orderBy($"ciudad", $"categoria")
  .show()

+--------+----------+---------------+
|  ciudad| categoria|total_facturado|
+--------+----------+---------------+
|    NULL|      NULL|         4100.0|
|    NULL|   Oficina|          730.0|
|    NULL|Tecnología|         3370.0|
|  Madrid|      NULL|         3350.0|
|  Madrid|   Oficina|          250.0|
|  Madrid|Tecnología|         3100.0|
|Valencia|      NULL|          750.0|
|Valencia|   Oficina|          480.0|
|Valencia|Tecnología|          270.0|
+--------+----------+---------------+



11. pivot() — tabla dinámica

In [71]:
val dfPivot = Seq(
  (1, "Madrid", "Tecnología", 2400.0),
  (2, "Madrid", "Tecnología", 700.0),
  (3, "Madrid", "Oficina", 250.0),
  (4, "Valencia", "Oficina", 480.0),
  (5, "Valencia", "Tecnología", 270.0)
).toDF("id", "ciudad", "categoria", "total")

dfPivot.show()

+---+--------+----------+------+
| id|  ciudad| categoria| total|
+---+--------+----------+------+
|  1|  Madrid|Tecnología|2400.0|
|  2|  Madrid|Tecnología| 700.0|
|  3|  Madrid|   Oficina| 250.0|
|  4|Valencia|   Oficina| 480.0|
|  5|Valencia|Tecnología| 270.0|
+---+--------+----------+------+



dfPivot: org.apache.spark.sql.package.DataFrame = [id: int, ciudad: string ... 2 more fields]

Una empresa registra sus ventas indicando la ciudad, la categoría del producto y el importe total de cada venta. La dirección quiere transformar estos datos en una estructura similar a una tabla dinámica, donde cada ciudad aparezca en una fila y cada categoría de producto se convierta en una columna independiente.

Se necesita calcular:

1. El total facturado por ciudad.
2. El total facturado por cada categoría dentro de cada ciudad.
3. Una tabla final donde las categorías, como Oficina y Tecnología, aparezcan como columnas.

Para resolverlo, utiliza groupBy("ciudad") para agrupar las ventas por ciudad y pivot("categoria") para convertir los valores de la columna categoria en nuevas columnas. Después, aplica sum("total") para calcular el total facturado en cada combinación ciudad-categoría.

In [72]:
dfPivot
  .groupBy("ciudad")
  .pivot("categoria")
  .agg(sum("total"))
  .orderBy("ciudad")
  .show()

+--------+-------+----------+
|  ciudad|Oficina|Tecnología|
+--------+-------+----------+
|  Madrid|  250.0|    3100.0|
|Valencia|  480.0|     270.0|
+--------+-------+----------+



💻 Parte 2 — Práctica

In [73]:
import $ivy.`org.apache.spark::spark-core:4.1.1`
import $ivy.`org.apache.spark::spark-sql:4.1.1`

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.expressions.Window

val spark = SparkSession.builder()
  .appName("Dia15_Sesion1_Agregaciones")
  .master("local[*]")
  .getOrCreate()

import spark.implicits._

spark.sparkContext.setLogLevel("ERROR")

println(s"✅ Spark ${spark.version} listo")

✅ Spark 4.1.1 listo


import $ivy.$
import $ivy.$
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.expressions.Window
spark: SparkSession = org.apache.spark.sql.classic.SparkSession@4f6e12b4
import spark.implicits._

In [74]:
val ventasMayo = Seq(
  (1, "2026-05-01", "Madrid", "Tecnología", "Laptop", 1200.0, 2),
  (2, "2026-05-01", "Barcelona", "Tecnología", "Monitor", 350.0, 1),
  (3, "2026-05-01", "Valencia", "Oficina", "Silla", 120.0, 4),
  (4, "2026-05-02", "Madrid", "Tecnología", "Teclado", 45.0, 6),
  (5, "2026-05-02", "Sevilla", "Oficina", "Mesa", 250.0, 2),
  (6, "2026-05-02", "Bilbao", "Tecnología", "Tablet", 600.0, 1),
  (7, "2026-05-03", "Madrid", "Tecnología", "Monitor", 350.0, 2),
  (8, "2026-05-03", "Barcelona", "Oficina", "Silla", 120.0, 3),
  (9, "2026-05-03", "Valencia", "Tecnología", "Ratón", 25.0, 10),
  (10, "2026-05-04", "Madrid", "Oficina", "Mesa", 250.0, 1),
  (11, "2026-05-04", "Sevilla", "Tecnología", "Laptop", 1200.0, 1),
  (12, "2026-05-04", "Bilbao", "Tecnología", "Teclado", 45.0, 5),
  (13, "2026-05-05", "Madrid", "Tecnología", "Tablet", 600.0, 2),
  (14, "2026-05-05", "Barcelona", "Tecnología", "Laptop", 1200.0, 1),
  (15, "2026-05-05", "Valencia", "Oficina", "Archivador", 80.0, 4),
  (16, "2026-05-06", "Madrid", "Tecnología", "Ratón", 25.0, 12),
  (17, "2026-05-06", "Sevilla", "Tecnología", "Monitor", 350.0, 1),
  (18, "2026-05-06", "Bilbao", "Oficina", "Silla", 120.0, 2),
  (19, "2026-05-07", "Madrid", "Oficina", "Silla", 120.0, 5),
  (20, "2026-05-07", "Barcelona", "Tecnología", "Tablet", 600.0, 2),
  (21, "2026-05-07", "Valencia", "Tecnología", "Teclado", 45.0, 7),
  (22, "2026-05-08", "Madrid", "Tecnología", "Laptop", 1200.0, 1),
  (23, "2026-05-08", "Sevilla", "Oficina", "Mesa", 250.0, 1),
  (24, "2026-05-08", "Bilbao", "Tecnología", "Monitor", 350.0, 2),
  (25, "2026-05-09", "Madrid", "Tecnología", "Monitor", 350.0, 3),
  (26, "2026-05-09", "Barcelona", "Tecnología", "Ratón", 25.0, 15),
  (27, "2026-05-09", "Valencia", "Oficina", "Mesa", 250.0, 2),
  (28, "2026-05-10", "Madrid", "Oficina", "Archivador", 80.0, 6),
  (29, "2026-05-10", "Sevilla", "Tecnología", "Tablet", 600.0, 1),
  (30, "2026-05-10", "Bilbao", "Tecnología", "Laptop", 1200.0, 1),
  (31, "2026-05-11", "Madrid", "Tecnología", "Teclado", 45.0, 8),
  (32, "2026-05-11", "Barcelona", "Oficina", "Mesa", 250.0, 1),
  (33, "2026-05-11", "Valencia", "Tecnología", "Monitor", 350.0, 1),
  (34, "2026-05-12", "Madrid", "Tecnología", "Laptop", 1200.0, 3),
  (35, "2026-05-12", "Sevilla", "Oficina", "Silla", 120.0, 4),
  (36, "2026-05-12", "Bilbao", "Tecnología", "Ratón", 25.0, 9),
  (37, "2026-05-13", "Madrid", "Oficina", "Mesa", 250.0, 2),
  (38, "2026-05-13", "Barcelona", "Tecnología", "Monitor", 350.0, 2),
  (39, "2026-05-13", "Valencia", "Tecnología", "Tablet", 600.0, 1),
  (40, "2026-05-14", "Madrid", "Tecnología", "Laptop", 1200.0, 2),
  (41, "2026-05-14", "Sevilla", "Tecnología", "Teclado", 45.0, 6),
  (42, "2026-05-14", "Bilbao", "Oficina", "Archivador", 80.0, 3),
  (43, "2026-05-15", "Madrid", "Tecnología", "Tablet", 600.0, 1),
  (44, "2026-05-15", "Barcelona", "Oficina", "Silla", 120.0, 2),
  (45, "2026-05-15", "Valencia", "Tecnología", "Laptop", 1200.0, 1)
).toDF("id", "fecha", "ciudad", "categoria", "producto", "precio_unitario", "cantidad")

val ventas = ventasMayo
  .withColumn("total", $"precio_unitario" * $"cantidad")

ventas.show(10)

+---+----------+---------+----------+--------+---------------+--------+------+
| id|     fecha|   ciudad| categoria|producto|precio_unitario|cantidad| total|
+---+----------+---------+----------+--------+---------------+--------+------+
|  1|2026-05-01|   Madrid|Tecnología|  Laptop|         1200.0|       2|2400.0|
|  2|2026-05-01|Barcelona|Tecnología| Monitor|          350.0|       1| 350.0|
|  3|2026-05-01| Valencia|   Oficina|   Silla|          120.0|       4| 480.0|
|  4|2026-05-02|   Madrid|Tecnología| Teclado|           45.0|       6| 270.0|
|  5|2026-05-02|  Sevilla|   Oficina|    Mesa|          250.0|       2| 500.0|
|  6|2026-05-02|   Bilbao|Tecnología|  Tablet|          600.0|       1| 600.0|
|  7|2026-05-03|   Madrid|Tecnología| Monitor|          350.0|       2| 700.0|
|  8|2026-05-03|Barcelona|   Oficina|   Silla|          120.0|       3| 360.0|
|  9|2026-05-03| Valencia|Tecnología|   Ratón|           25.0|      10| 250.0|
| 10|2026-05-04|   Madrid|   Oficina|    Mesa|      

ventasMayo: org.apache.spark.sql.package.DataFrame = [id: int, fecha: string ... 5 more fields]
ventas: org.apache.spark.sql.package.DataFrame = [id: int, fecha: string ... 6 more fields]

In [75]:
ventas.show(5)

+---+----------+---------+----------+--------+---------------+--------+------+
| id|     fecha|   ciudad| categoria|producto|precio_unitario|cantidad| total|
+---+----------+---------+----------+--------+---------------+--------+------+
|  1|2026-05-01|   Madrid|Tecnología|  Laptop|         1200.0|       2|2400.0|
|  2|2026-05-01|Barcelona|Tecnología| Monitor|          350.0|       1| 350.0|
|  3|2026-05-01| Valencia|   Oficina|   Silla|          120.0|       4| 480.0|
|  4|2026-05-02|   Madrid|Tecnología| Teclado|           45.0|       6| 270.0|
|  5|2026-05-02|  Sevilla|   Oficina|    Mesa|          250.0|       2| 500.0|
+---+----------+---------+----------+--------+---------------+--------+------+
only showing top 5 rows


In [76]:
ventas.printSchema()

root
 |-- id: integer (nullable = false)
 |-- fecha: string (nullable = true)
 |-- ciudad: string (nullable = true)
 |-- categoria: string (nullable = true)
 |-- producto: string (nullable = true)
 |-- precio_unitario: double (nullable = false)
 |-- cantidad: integer (nullable = false)
 |-- total: double (nullable = false)



In [77]:
ventas.count()

res77: Long = 45L

In [78]:
ventas.describe("precio_unitario", "cantidad", "total").show()

+-------+------------------+------------------+-----------------+
|summary|   precio_unitario|          cantidad|            total|
+-------+------------------+------------------+-----------------+
|  count|                45|                45|               45|
|   mean| 409.6666666666667|3.3777777777777778|            704.0|
| stddev|412.26977046148266| 3.214236010518383|670.8166664596222|
|    min|              25.0|                 1|            225.0|
|    max|            1200.0|                15|           3600.0|
+-------+------------------+------------------+-----------------+



Preguntas rápidas
1. ¿Cuántas filas tiene el DataFrame? 45 
2. ¿Qué tipo de dato tiene la columna total? double
3. ¿Cuál es el valor máximo de cantidad? 15
4. ¿Cuál es el valor máximo de total? 225

In [79]:
val ventasPorCiudad = ventas
  .groupBy("ciudad")
  .count()
  .orderBy($"count".desc)

ventasPorCiudad.show()

+---------+-----+
|   ciudad|count|
+---------+-----+
|   Madrid|   15|
|Barcelona|    8|
| Valencia|    8|
|  Sevilla|    7|
|   Bilbao|    7|
+---------+-----+



ventasPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, count: bigint]

In [80]:
val ventasPorCategoria = ventas
  .groupBy("categoria")
  .count()
  .orderBy($"count".desc)

ventasPorCategoria.show()

+----------+-----+
| categoria|count|
+----------+-----+
|Tecnología|   30|
|   Oficina|   15|
+----------+-----+



ventasPorCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [categoria: string, count: bigint]

In [81]:
val ventasPorProducto = ventas
  .groupBy("producto")
  .count()
  .orderBy($"count".desc)

ventasPorProducto.show()

+----------+-----+
|  producto|count|
+----------+-----+
|    Laptop|    8|
|   Monitor|    7|
|     Silla|    6|
|      Mesa|    6|
|    Tablet|    6|
|   Teclado|    5|
|     Ratón|    4|
|Archivador|    3|
+----------+-----+



ventasPorProducto: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [producto: string, count: bigint]

Preguntas
1. ¿Qué ciudad tiene más registros de venta? Madrid
2. ¿Qué categoría aparece más veces? tecnologia
3. ¿Qué producto se vendió en más operaciones? laptop


🧪 Bloque 3 — sum, avg, min, max

In [82]:
val facturacionPorCiudad = ventas
  .groupBy("ciudad")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"total_facturado".desc)

facturacionPorCiudad.show()

+---------+---------------+
|   ciudad|total_facturado|
+---------+---------------+
|   Madrid|        15910.0|
|Barcelona|         4675.0|
| Valencia|         4015.0|
|  Sevilla|         3650.0|
|   Bilbao|         3430.0|
+---------+---------------+



facturacionPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, total_facturado: double]

In [83]:
val facturacionPorCategoria = ventas
  .groupBy("categoria")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"total_facturado".desc)

facturacionPorCategoria.show()

+----------+---------------+
| categoria|total_facturado|
+----------+---------------+
|Tecnología|        25990.0|
|   Oficina|         5690.0|
+----------+---------------+



facturacionPorCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [categoria: string, total_facturado: double]

In [84]:
val ticketMedioPorCiudad = ventas
  .groupBy("ciudad")
  .agg(
    avg("total").as("ticket_medio")
  )
  .orderBy($"ticket_medio".desc)

ticketMedioPorCiudad.show()

+---------+------------------+
|   ciudad|      ticket_medio|
+---------+------------------+
|   Madrid|1060.6666666666667|
|Barcelona|           584.375|
|  Sevilla| 521.4285714285714|
| Valencia|           501.875|
|   Bilbao|             490.0|
+---------+------------------+



ticketMedioPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, ticket_medio: double]

In [85]:
val minMaxPorCiudad = ventas
  .groupBy("ciudad")
  .agg(
    min("total").as("venta_minima"),
    max("total").as("venta_maxima")
  )
  .orderBy($"venta_maxima".desc)

minMaxPorCiudad.show()

+---------+------------+------------+
|   ciudad|venta_minima|venta_maxima|
+---------+------------+------------+
|   Madrid|       250.0|      3600.0|
|Barcelona|       240.0|      1200.0|
| Valencia|       250.0|      1200.0|
|  Sevilla|       250.0|      1200.0|
|   Bilbao|       225.0|      1200.0|
+---------+------------+------------+



minMaxPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, venta_minima: double ... 1 more field]

1. ¿Qué ciudad factura más? Madrid
2. ¿Qué categoría tiene mayor facturación total? Tecnología
3. ¿Qué ciudad tiene el ticket medio más alto? Madrid
4. ¿En qué ciudad se encuentra la venta máxima? Madrid

🧪 Bloque 4 — agg() con varias métricas

In [86]:
val resumenCiudad = ventas
  .groupBy("ciudad")
  .agg(
    count("id").as("num_ventas"),
    sum("cantidad").as("unidades_vendidas"),
    sum("total").as("total_facturado"),
    avg("total").as("ticket_medio"),
    max("total").as("venta_maxima")
  )
  .orderBy($"total_facturado".desc)

resumenCiudad.show()

+---------+----------+-----------------+---------------+------------------+------------+
|   ciudad|num_ventas|unidades_vendidas|total_facturado|      ticket_medio|venta_maxima|
+---------+----------+-----------------+---------------+------------------+------------+
|   Madrid|        15|               56|        15910.0|1060.6666666666667|      3600.0|
|Barcelona|         8|               27|         4675.0|           584.375|      1200.0|
| Valencia|         8|               30|         4015.0|           501.875|      1200.0|
|  Sevilla|         7|               16|         3650.0| 521.4285714285714|      1200.0|
|   Bilbao|         7|               23|         3430.0|             490.0|      1200.0|
+---------+----------+-----------------+---------------+------------------+------------+



resumenCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, num_ventas: bigint ... 4 more fields]

In [87]:
val resumenProducto = ventas
  .groupBy("producto")
  .agg(
    count("id").as("num_ventas"),
    sum("cantidad").as("unidades_vendidas"),
    sum("total").as("total_facturado"),
    avg("total").as("ticket_medio"),
    max("total").as("venta_maxima")
  )
  .orderBy($"total_facturado".desc)

resumenProducto.show()

+----------+----------+-----------------+---------------+-----------------+------------+
|  producto|num_ventas|unidades_vendidas|total_facturado|     ticket_medio|venta_maxima|
+----------+----------+-----------------+---------------+-----------------+------------+
|    Laptop|         8|               12|        14400.0|           1800.0|      3600.0|
|    Tablet|         6|                8|         4800.0|            800.0|      1200.0|
|   Monitor|         7|               12|         4200.0|            600.0|      1050.0|
|     Silla|         6|               20|         2400.0|            400.0|       600.0|
|      Mesa|         6|                9|         2250.0|            375.0|       500.0|
|   Teclado|         5|               32|         1440.0|            288.0|       360.0|
|     Ratón|         4|               46|         1150.0|            287.5|       375.0|
|Archivador|         3|               13|         1040.0|346.6666666666667|       480.0|
+----------+---------

resumenProducto: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [producto: string, num_ventas: bigint ... 4 more fields]

In [88]:
val resumenCiudadCategoria = ventas
  .groupBy("ciudad", "categoria")
  .agg(
    count("id").as("num_ventas"),
    sum("total").as("total_facturado")
  )
  .orderBy($"ciudad", $"total_facturado".desc)

resumenCiudadCategoria.show()

+---------+----------+----------+---------------+
|   ciudad| categoria|num_ventas|total_facturado|
+---------+----------+----------+---------------+
|Barcelona|Tecnología|         5|         3825.0|
|Barcelona|   Oficina|         3|          850.0|
|   Bilbao|Tecnología|         5|         2950.0|
|   Bilbao|   Oficina|         2|          480.0|
|   Madrid|Tecnología|        11|        14080.0|
|   Madrid|   Oficina|         4|         1830.0|
|  Sevilla|Tecnología|         4|         2420.0|
|  Sevilla|   Oficina|         3|         1230.0|
| Valencia|Tecnología|         5|         2715.0|
| Valencia|   Oficina|         3|         1300.0|
+---------+----------+----------+---------------+



resumenCiudadCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, categoria: string ... 2 more fields]

Preguntas
1. ¿Qué producto genera más facturación? Laptop
2. ¿Qué producto tiene más unidades vendidas? Ratón
3. ¿Qué combinación ciudad-categoría factura más? Madrid - Tecnología
4. ¿Por qué agg() es más práctico que hacer varias consultas separadas? 

🧪 Bloque 5 — Ranking con funciones de ventana

In [89]:
val ventanaCiudad = Window
  .partitionBy("ciudad")
  .orderBy($"total".desc)

val rankingVentasCiudad = ventas
  .withColumn("ranking_ciudad", row_number().over(ventanaCiudad))
  .orderBy($"ciudad", $"ranking_ciudad")

rankingVentasCiudad.show(50)

+---+----------+---------+----------+----------+---------------+--------+------+--------------+
| id|     fecha|   ciudad| categoria|  producto|precio_unitario|cantidad| total|ranking_ciudad|
+---+----------+---------+----------+----------+---------------+--------+------+--------------+
| 14|2026-05-05|Barcelona|Tecnología|    Laptop|         1200.0|       1|1200.0|             1|
| 20|2026-05-07|Barcelona|Tecnología|    Tablet|          600.0|       2|1200.0|             2|
| 38|2026-05-13|Barcelona|Tecnología|   Monitor|          350.0|       2| 700.0|             3|
| 26|2026-05-09|Barcelona|Tecnología|     Ratón|           25.0|      15| 375.0|             4|
|  8|2026-05-03|Barcelona|   Oficina|     Silla|          120.0|       3| 360.0|             5|
|  2|2026-05-01|Barcelona|Tecnología|   Monitor|          350.0|       1| 350.0|             6|
| 32|2026-05-11|Barcelona|   Oficina|      Mesa|          250.0|       1| 250.0|             7|
| 44|2026-05-15|Barcelona|   Oficina|   

ventanaCiudad: org.apache.spark.sql.expressions.WindowSpec = org.apache.spark.sql.expressions.WindowSpec@753549fb
rankingVentasCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [id: int, fecha: string ... 7 more fields]

In [90]:
val topVentaPorCiudad = rankingVentasCiudad
  .filter($"ranking_ciudad" === 1)
  .select("ciudad", "producto", "total", "ranking_ciudad")
  .orderBy($"total".desc)

topVentaPorCiudad.show()

+---------+--------+------+--------------+
|   ciudad|producto| total|ranking_ciudad|
+---------+--------+------+--------------+
|   Madrid|  Laptop|3600.0|             1|
|Barcelona|  Laptop|1200.0|             1|
|   Bilbao|  Laptop|1200.0|             1|
|  Sevilla|  Laptop|1200.0|             1|
| Valencia|  Laptop|1200.0|             1|
+---------+--------+------+--------------+



topVentaPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, producto: string ... 2 more fields]

In [91]:
val top3VentasPorCiudad = rankingVentasCiudad
  .filter($"ranking_ciudad" <= 3)
  .select("ciudad", "producto", "total", "ranking_ciudad")
  .orderBy($"ciudad", $"ranking_ciudad")

top3VentasPorCiudad.show(50)

+---------+--------+------+--------------+
|   ciudad|producto| total|ranking_ciudad|
+---------+--------+------+--------------+
|Barcelona|  Laptop|1200.0|             1|
|Barcelona|  Tablet|1200.0|             2|
|Barcelona| Monitor| 700.0|             3|
|   Bilbao|  Laptop|1200.0|             1|
|   Bilbao| Monitor| 700.0|             2|
|   Bilbao|  Tablet| 600.0|             3|
|   Madrid|  Laptop|3600.0|             1|
|   Madrid|  Laptop|2400.0|             2|
|   Madrid|  Laptop|2400.0|             3|
|  Sevilla|  Laptop|1200.0|             1|
|  Sevilla|  Tablet| 600.0|             2|
|  Sevilla|    Mesa| 500.0|             3|
| Valencia|  Laptop|1200.0|             1|
| Valencia|  Tablet| 600.0|             2|
| Valencia|    Mesa| 500.0|             3|
+---------+--------+------+--------------+



top3VentasPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, producto: string ... 2 more fields]

Preguntas
1. ¿Qué venta ocupa el primer puesto en Madrid? Laptop
2. ¿Qué producto aparece más veces en los primeros puestos? Laptop
3. ¿Qué diferencia hay entre ordenar todo el DataFrame y ordenar dentro de cada ciudad? 

🧪 Bloque 6 — rollup()

In [92]:
val rollupCiudadCategoria = ventas
  .rollup("ciudad", "categoria")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"ciudad", $"categoria")

rollupCiudadCategoria.show(50)

+---------+----------+---------------+
|   ciudad| categoria|total_facturado|
+---------+----------+---------------+
|     NULL|      NULL|        31680.0|
|Barcelona|      NULL|         4675.0|
|Barcelona|   Oficina|          850.0|
|Barcelona|Tecnología|         3825.0|
|   Bilbao|      NULL|         3430.0|
|   Bilbao|   Oficina|          480.0|
|   Bilbao|Tecnología|         2950.0|
|   Madrid|      NULL|        15910.0|
|   Madrid|   Oficina|         1830.0|
|   Madrid|Tecnología|        14080.0|
|  Sevilla|      NULL|         3650.0|
|  Sevilla|   Oficina|         1230.0|
|  Sevilla|Tecnología|         2420.0|
| Valencia|      NULL|         4015.0|
| Valencia|   Oficina|         1300.0|
| Valencia|Tecnología|         2715.0|
+---------+----------+---------------+



rollupCiudadCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, categoria: string ... 1 more field]

Preguntas
1. ¿Qué representan las filas donde categoria aparece como null? NO hay informacion.
2. ¿Qué representa la fila donde ciudad y categoria aparecen como null? El total de todo el dataset es 31680.
3. ¿Por qué puede ser útil rollup() para crear informes? Porque trabaja como tabla dinamica.

🧪 Bloque 7 — cube()

In [93]:
val cubeCiudadCategoria = ventas
  .cube("ciudad", "categoria")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"ciudad", $"categoria")

cubeCiudadCategoria.show(50)

+---------+----------+---------------+
|   ciudad| categoria|total_facturado|
+---------+----------+---------------+
|     NULL|      NULL|        31680.0|
|     NULL|   Oficina|         5690.0|
|     NULL|Tecnología|        25990.0|
|Barcelona|      NULL|         4675.0|
|Barcelona|   Oficina|          850.0|
|Barcelona|Tecnología|         3825.0|
|   Bilbao|      NULL|         3430.0|
|   Bilbao|   Oficina|          480.0|
|   Bilbao|Tecnología|         2950.0|
|   Madrid|      NULL|        15910.0|
|   Madrid|   Oficina|         1830.0|
|   Madrid|Tecnología|        14080.0|
|  Sevilla|      NULL|         3650.0|
|  Sevilla|   Oficina|         1230.0|
|  Sevilla|Tecnología|         2420.0|
| Valencia|      NULL|         4015.0|
| Valencia|   Oficina|         1300.0|
| Valencia|Tecnología|         2715.0|
+---------+----------+---------------+



cubeCiudadCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, categoria: string ... 1 more field]

Preguntas
1. ¿Qué diferencia observas entre rollup() y cube()? Cube realiza mas combinaciones que rollup
2. ¿Qué representan las filas donde ciudad es null pero categoria tiene valor? Es el total por categoria.
3. ¿Qué representa la fila donde ambas columnas son null? Es el total del data set y es 31680.

🧪 Bloque 8 — pivot()

In [94]:
val pivotCiudadProducto = ventas
  .groupBy("ciudad")
  .pivot("producto")
  .agg(sum("total"))
  .orderBy("ciudad")

pivotCiudadProducto.show()

+---------+----------+------+-----+-------+-----+-----+------+-------+
|   ciudad|Archivador|Laptop| Mesa|Monitor|Ratón|Silla|Tablet|Teclado|
+---------+----------+------+-----+-------+-----+-----+------+-------+
|Barcelona|      NULL|1200.0|250.0| 1050.0|375.0|600.0|1200.0|   NULL|
|   Bilbao|     240.0|1200.0| NULL|  700.0|225.0|240.0| 600.0|  225.0|
|   Madrid|     480.0|9600.0|750.0| 1750.0|300.0|600.0|1800.0|  630.0|
|  Sevilla|      NULL|1200.0|750.0|  350.0| NULL|480.0| 600.0|  270.0|
| Valencia|     320.0|1200.0|500.0|  350.0|250.0|480.0| 600.0|  315.0|
+---------+----------+------+-----+-------+-----+-----+------+-------+



pivotCiudadProducto: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, Archivador: double ... 7 more fields]

Preguntas
1. ¿Qué ciudad factura más en Tecnología? No se sabe, ya que no han estructura por la categoria, sino por el producto.
2. ¿Qué ciudad factura más en Oficina? No se sabe, ya que no han estructurado por la categoria, sino por el producto.
3. ¿Para qué tipo de informe sería útil un pivot()? Para realizar combinaciones entre dos columnas. asi como las tablas dinamicas.
4. ¿Qué problema podría aparecer si una columna tiene demasiados valores distintos? Ninguno, es la informaciòn que se muestra en el cruce de dos campos. Si hay NULL, es que no hay la coincidencia.

🏢 Caso de Estudio Propuesto — MarketNova Analytics

MarketNova es una cadena de tiendas online que vende productos de distintas categorías. Durante el mes de mayo ha registrado ventas en varias ciudades españolas. El equipo directivo quiere un primer informe para responder preguntas como:

    * Qué ciudad está generando más ingresos
    * Qué categoría es más rentable
    * Qué productos lideran las ventas
    * Cuáles son las mejores ventas dentro de cada ciudad
    * Cómo se distribuye la facturación por ciudad y categoría

Tu tarea como analista de datos es construir un análisis con DataFrames usando únicamente las herramientas vistas en esta sesión.

In [95]:
import $ivy.`org.apache.spark::spark-core:4.1.1`
import $ivy.`org.apache.spark::spark-sql:4.1.1`

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.expressions.Window

val spark = SparkSession.builder()
  .appName("MarketNova_Analytics")
  .master("local[*]")
  .getOrCreate()

import spark.implicits._
spark.sparkContext.setLogLevel("ERROR")

import $ivy.$
import $ivy.$
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.expressions.Window
spark: SparkSession = org.apache.spark.sql.classic.SparkSession@4f6e12b4
import spark.implicits._

In [96]:
val marketNovaRaw = Seq(
  (1, "2026-05-01", "Madrid", "Electrónica", "Portátil", 950.0, 2),
  (2, "2026-05-01", "Madrid", "Hogar", "Aspiradora", 180.0, 1),
  (3, "2026-05-02", "Barcelona", "Electrónica", "Tablet", 420.0, 3),
  (4, "2026-05-02", "Valencia", "Deporte", "Bicicleta", 300.0, 2),
  (5, "2026-05-03", "Sevilla", "Hogar", "Cafetera", 90.0, 4),
  (6, "2026-05-03", "Bilbao", "Electrónica", "Auriculares", 75.0, 6),
  (7, "2026-05-04", "Madrid", "Deporte", "Cinta correr", 650.0, 1),
  (8, "2026-05-04", "Barcelona", "Hogar", "Microondas", 140.0, 2),
  (9, "2026-05-05", "Valencia", "Electrónica", "Monitor", 280.0, 2),
  (10, "2026-05-05", "Sevilla", "Deporte", "Mancuernas", 45.0, 8),
  (11, "2026-05-06", "Bilbao", "Hogar", "Batidora", 60.0, 5),
  (12, "2026-05-06", "Madrid", "Electrónica", "Smartphone", 700.0, 2),
  (13, "2026-05-07", "Barcelona", "Deporte", "Bicicleta", 300.0, 1),
  (14, "2026-05-07", "Valencia", "Hogar", "Cafetera", 90.0, 3),
  (15, "2026-05-08", "Sevilla", "Electrónica", "Tablet", 420.0, 2),
  (16, "2026-05-08", "Bilbao", "Deporte", "Balón", 25.0, 10),
  (17, "2026-05-09", "Madrid", "Hogar", "Microondas", 140.0, 3),
  (18, "2026-05-09", "Barcelona", "Electrónica", "Portátil", 950.0, 1),
  (19, "2026-05-10", "Valencia", "Deporte", "Cinta correr", 650.0, 1),
  (20, "2026-05-10", "Sevilla", "Hogar", "Aspiradora", 180.0, 2),
  (21, "2026-05-11", "Bilbao", "Electrónica", "Smartphone", 700.0, 1),
  (22, "2026-05-11", "Madrid", "Deporte", "Bicicleta", 300.0, 2),
  (23, "2026-05-12", "Barcelona", "Hogar", "Cafetera", 90.0, 5),
  (24, "2026-05-12", "Valencia", "Electrónica", "Auriculares", 75.0, 7),
  (25, "2026-05-13", "Sevilla", "Deporte", "Cinta correr", 650.0, 1),
  (26, "2026-05-13", "Bilbao", "Hogar", "Microondas", 140.0, 2),
  (27, "2026-05-14", "Madrid", "Electrónica", "Monitor", 280.0, 4),
  (28, "2026-05-14", "Barcelona", "Deporte", "Mancuernas", 45.0, 6),
  (29, "2026-05-15", "Valencia", "Hogar", "Aspiradora", 180.0, 2),
  (30, "2026-05-15", "Sevilla", "Electrónica", "Smartphone", 700.0, 1),
  (31, "2026-05-16", "Bilbao", "Deporte", "Bicicleta", 300.0, 2),
  (32, "2026-05-16", "Madrid", "Hogar", "Batidora", 60.0, 5),
  (33, "2026-05-17", "Barcelona", "Electrónica", "Monitor", 280.0, 3),
  (34, "2026-05-17", "Valencia", "Deporte", "Balón", 25.0, 12),
  (35, "2026-05-18", "Sevilla", "Hogar", "Microondas", 140.0, 2),
  (36, "2026-05-18", "Bilbao", "Electrónica", "Portátil", 950.0, 1)
).toDF("id", "fecha", "ciudad", "categoria", "producto", "precio_unitario", "cantidad")

val marketNova = marketNovaRaw
  .withColumn("total", $"precio_unitario" * $"cantidad")

marketNova.show(10)

+---+----------+---------+-----------+------------+---------------+--------+------+
| id|     fecha|   ciudad|  categoria|    producto|precio_unitario|cantidad| total|
+---+----------+---------+-----------+------------+---------------+--------+------+
|  1|2026-05-01|   Madrid|Electrónica|    Portátil|          950.0|       2|1900.0|
|  2|2026-05-01|   Madrid|      Hogar|  Aspiradora|          180.0|       1| 180.0|
|  3|2026-05-02|Barcelona|Electrónica|      Tablet|          420.0|       3|1260.0|
|  4|2026-05-02| Valencia|    Deporte|   Bicicleta|          300.0|       2| 600.0|
|  5|2026-05-03|  Sevilla|      Hogar|    Cafetera|           90.0|       4| 360.0|
|  6|2026-05-03|   Bilbao|Electrónica| Auriculares|           75.0|       6| 450.0|
|  7|2026-05-04|   Madrid|    Deporte|Cinta correr|          650.0|       1| 650.0|
|  8|2026-05-04|Barcelona|      Hogar|  Microondas|          140.0|       2| 280.0|
|  9|2026-05-05| Valencia|Electrónica|     Monitor|          280.0|       2|

marketNovaRaw: org.apache.spark.sql.package.DataFrame = [id: int, fecha: string ... 5 more fields]
marketNova: org.apache.spark.sql.package.DataFrame = [id: int, fecha: string ... 6 more fields]

Pregunta 1 — Exploración inicial

Muestra las primeras 10 filas, el schema y las estadísticas de precio_unitario, cantidad y total.

In [97]:
marketNova.show(10)

+---+----------+---------+-----------+------------+---------------+--------+------+
| id|     fecha|   ciudad|  categoria|    producto|precio_unitario|cantidad| total|
+---+----------+---------+-----------+------------+---------------+--------+------+
|  1|2026-05-01|   Madrid|Electrónica|    Portátil|          950.0|       2|1900.0|
|  2|2026-05-01|   Madrid|      Hogar|  Aspiradora|          180.0|       1| 180.0|
|  3|2026-05-02|Barcelona|Electrónica|      Tablet|          420.0|       3|1260.0|
|  4|2026-05-02| Valencia|    Deporte|   Bicicleta|          300.0|       2| 600.0|
|  5|2026-05-03|  Sevilla|      Hogar|    Cafetera|           90.0|       4| 360.0|
|  6|2026-05-03|   Bilbao|Electrónica| Auriculares|           75.0|       6| 450.0|
|  7|2026-05-04|   Madrid|    Deporte|Cinta correr|          650.0|       1| 650.0|
|  8|2026-05-04|Barcelona|      Hogar|  Microondas|          140.0|       2| 280.0|
|  9|2026-05-05| Valencia|Electrónica|     Monitor|          280.0|       2|

In [98]:
marketNova.printSchema()

root
 |-- id: integer (nullable = false)
 |-- fecha: string (nullable = true)
 |-- ciudad: string (nullable = true)
 |-- categoria: string (nullable = true)
 |-- producto: string (nullable = true)
 |-- precio_unitario: double (nullable = false)
 |-- cantidad: integer (nullable = false)
 |-- total: double (nullable = false)



In [99]:
marketNova.describe("precio_unitario", "cantidad", "total").show()

+-------+------------------+------------------+------------------+
|summary|   precio_unitario|          cantidad|             total|
+-------+------------------+------------------+------------------+
|  count|                36|                36|                36|
|   mean| 321.1111111111111|              3.25| 590.6944444444445|
| stddev|288.07764605871483|2.6659225152173605|374.90980396771585|
|    min|              25.0|                 1|             180.0|
|    max|             950.0|                12|            1900.0|
+-------+------------------+------------------+------------------+



In [100]:
marketNova.count()

res100: Long = 36L

Pregunta 2 — Ventas por ciudad

Calcula cuántas ventas hay por ciudad y ordena de mayor a menor.

In [101]:
val ventasPorCiudad = marketNova
  .groupBy("ciudad")
  .count()
  .orderBy($"count".desc)

ventasPorCiudad.show()

+---------+-----+
|   ciudad|count|
+---------+-----+
|   Madrid|    8|
|Barcelona|    7|
| Valencia|    7|
|   Bilbao|    7|
|  Sevilla|    7|
+---------+-----+



ventasPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, count: bigint]

Pregunta 3 — Facturación por ciudad

Calcula la facturación total por ciudad.

In [102]:
val facturacionPorCiudad = marketNova
  .groupBy("ciudad")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"total_facturado".desc)

facturacionPorCiudad.show()

+---------+---------------+
|   ciudad|total_facturado|
+---------+---------------+
|   Madrid|         6570.0|
|Barcelona|         4350.0|
|  Sevilla|         3550.0|
|   Bilbao|         3530.0|
| Valencia|         3265.0|
+---------+---------------+



facturacionPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, total_facturado: double]

Pregunta 4 — Facturación por categoría

Calcula la facturación total por categoría.

In [103]:
val facturacionPorCategoria = marketNova
  .groupBy("categoria")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"total_facturado".desc)

facturacionPorCategoria.show()

+-----------+---------------+
|  categoria|total_facturado|
+-----------+---------------+
|Electrónica|        12195.0|
|    Deporte|         5230.0|
|      Hogar|         3840.0|
+-----------+---------------+



facturacionPorCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [categoria: string, total_facturado: double]

Pregunta 5 — Resumen por producto

Para cada producto calcula:
1. Número de ventas
2. Unidades vendidas
3. Total facturado
4. Ticket medio
5. Venta máxima

In [104]:
val resumenProducto = marketNova
  .groupBy("ciudad")
  .agg(
    count("id").as("num_ventas"),
    sum("cantidad").as("unidades_vendidas"),
    sum("total").as("total_facturado"),
    avg("total").as("ticket_medio"),
    max("total").as("venta_maxima")
  )
  .orderBy($"total_facturado".desc)

resumenProducto.show()

+---------+----------+-----------------+---------------+------------------+------------+
|   ciudad|num_ventas|unidades_vendidas|total_facturado|      ticket_medio|venta_maxima|
+---------+----------+-----------------+---------------+------------------+------------+
|   Madrid|         8|               20|         6570.0|            821.25|      1900.0|
|Barcelona|         7|               21|         4350.0| 621.4285714285714|      1260.0|
|  Sevilla|         7|               20|         3550.0|507.14285714285717|       840.0|
|   Bilbao|         7|               27|         3530.0| 504.2857142857143|       950.0|
| Valencia|         7|               29|         3265.0|466.42857142857144|       650.0|
+---------+----------+-----------------+---------------+------------------+------------+



resumenProducto: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, num_ventas: bigint ... 4 more fields]

Pregunta 6 — Resumen por ciudad y categoría

Calcula el total facturado por cada combinación de ciudad y categoría.

In [106]:
val resumenCiudadCategoria = marketNova
  .groupBy("ciudad", "categoria")
  .agg(
    count("id").as("num_ventas"),
    sum("total").as("total_facturado")
  )
  .orderBy($"ciudad", $"total_facturado".desc)

resumenCiudadCategoria.show()

+---------+-----------+----------+---------------+
|   ciudad|  categoria|num_ventas|total_facturado|
+---------+-----------+----------+---------------+
|Barcelona|Electrónica|         3|         3050.0|
|Barcelona|      Hogar|         2|          730.0|
|Barcelona|    Deporte|         2|          570.0|
|   Bilbao|Electrónica|         3|         2100.0|
|   Bilbao|    Deporte|         2|          850.0|
|   Bilbao|      Hogar|         2|          580.0|
|   Madrid|Electrónica|         3|         4420.0|
|   Madrid|    Deporte|         2|         1250.0|
|   Madrid|      Hogar|         3|          900.0|
|  Sevilla|Electrónica|         2|         1540.0|
|  Sevilla|    Deporte|         2|         1010.0|
|  Sevilla|      Hogar|         3|         1000.0|
| Valencia|    Deporte|         3|         1550.0|
| Valencia|Electrónica|         2|         1085.0|
| Valencia|      Hogar|         2|          630.0|
+---------+-----------+----------+---------------+



resumenCiudadCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, categoria: string ... 2 more fields]

Pregunta 7 — Ranking de ventas por ciudad

Crea un ranking de ventas dentro de cada ciudad, ordenando por total de mayor a menor.

In [107]:
val ventanaCiudad = Window
  .partitionBy("ciudad")
  .orderBy($"total".desc)

val rankingVentasCiudad = marketNova
  .withColumn("ranking_ciudad", row_number().over(ventanaCiudad))
  .orderBy($"ciudad", $"ranking_ciudad")

rankingVentasCiudad.show(50)

+---+----------+---------+-----------+------------+---------------+--------+------+--------------+
| id|     fecha|   ciudad|  categoria|    producto|precio_unitario|cantidad| total|ranking_ciudad|
+---+----------+---------+-----------+------------+---------------+--------+------+--------------+
|  3|2026-05-02|Barcelona|Electrónica|      Tablet|          420.0|       3|1260.0|             1|
| 18|2026-05-09|Barcelona|Electrónica|    Portátil|          950.0|       1| 950.0|             2|
| 33|2026-05-17|Barcelona|Electrónica|     Monitor|          280.0|       3| 840.0|             3|
| 23|2026-05-12|Barcelona|      Hogar|    Cafetera|           90.0|       5| 450.0|             4|
| 13|2026-05-07|Barcelona|    Deporte|   Bicicleta|          300.0|       1| 300.0|             5|
|  8|2026-05-04|Barcelona|      Hogar|  Microondas|          140.0|       2| 280.0|             6|
| 28|2026-05-14|Barcelona|    Deporte|  Mancuernas|           45.0|       6| 270.0|             7|
| 36|2026-

ventanaCiudad: org.apache.spark.sql.expressions.WindowSpec = org.apache.spark.sql.expressions.WindowSpec@6511e648
rankingVentasCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [id: int, fecha: string ... 7 more fields]

Pregunta 8 — Mejor venta de cada ciudad

A partir del ranking anterior, muestra solo la venta número 1 de cada ciudad.

In [108]:
val topVentaPorCiudad = rankingVentasCiudad
  .filter($"ranking_ciudad" === 1)
  .select("ciudad", "producto", "total", "ranking_ciudad")
  .orderBy($"total".desc)

topVentaPorCiudad.show()

+---------+------------+------+--------------+
|   ciudad|    producto| total|ranking_ciudad|
+---------+------------+------+--------------+
|   Madrid|    Portátil|1900.0|             1|
|Barcelona|      Tablet|1260.0|             1|
|   Bilbao|    Portátil| 950.0|             1|
|  Sevilla|      Tablet| 840.0|             1|
| Valencia|Cinta correr| 650.0|             1|
+---------+------------+------+--------------+



topVentaPorCiudad: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, producto: string ... 2 more fields]

Pregunta 9 — Subtotales con rollup

Usa rollup() para obtener subtotales por ciudad y categoría.


In [109]:
val rollupCiudadCategoria = marketNova
  .rollup("ciudad", "categoria")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"ciudad", $"categoria")

rollupCiudadCategoria.show(50)

+---------+-----------+---------------+
|   ciudad|  categoria|total_facturado|
+---------+-----------+---------------+
|     NULL|       NULL|        21265.0|
|Barcelona|       NULL|         4350.0|
|Barcelona|    Deporte|          570.0|
|Barcelona|Electrónica|         3050.0|
|Barcelona|      Hogar|          730.0|
|   Bilbao|       NULL|         3530.0|
|   Bilbao|    Deporte|          850.0|
|   Bilbao|Electrónica|         2100.0|
|   Bilbao|      Hogar|          580.0|
|   Madrid|       NULL|         6570.0|
|   Madrid|    Deporte|         1250.0|
|   Madrid|Electrónica|         4420.0|
|   Madrid|      Hogar|          900.0|
|  Sevilla|       NULL|         3550.0|
|  Sevilla|    Deporte|         1010.0|
|  Sevilla|Electrónica|         1540.0|
|  Sevilla|      Hogar|         1000.0|
| Valencia|       NULL|         3265.0|
| Valencia|    Deporte|         1550.0|
| Valencia|Electrónica|         1085.0|
| Valencia|      Hogar|          630.0|
+---------+-----------+---------------+


rollupCiudadCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, categoria: string ... 1 more field]

Pregunta 10 — Análisis multidimensional con cube

Usa cube() para obtener totales por ciudad, por categoría, por combinación ciudad-categoría y total general.

In [110]:
val cubeCiudadCategoria = marketNova
  .cube("ciudad", "categoria")
  .agg(
    sum("total").as("total_facturado")
  )
  .orderBy($"ciudad", $"categoria")

cubeCiudadCategoria.show(50)

+---------+-----------+---------------+
|   ciudad|  categoria|total_facturado|
+---------+-----------+---------------+
|     NULL|       NULL|        21265.0|
|     NULL|    Deporte|         5230.0|
|     NULL|Electrónica|        12195.0|
|     NULL|      Hogar|         3840.0|
|Barcelona|       NULL|         4350.0|
|Barcelona|    Deporte|          570.0|
|Barcelona|Electrónica|         3050.0|
|Barcelona|      Hogar|          730.0|
|   Bilbao|       NULL|         3530.0|
|   Bilbao|    Deporte|          850.0|
|   Bilbao|Electrónica|         2100.0|
|   Bilbao|      Hogar|          580.0|
|   Madrid|       NULL|         6570.0|
|   Madrid|    Deporte|         1250.0|
|   Madrid|Electrónica|         4420.0|
|   Madrid|      Hogar|          900.0|
|  Sevilla|       NULL|         3550.0|
|  Sevilla|    Deporte|         1010.0|
|  Sevilla|Electrónica|         1540.0|
|  Sevilla|      Hogar|         1000.0|
| Valencia|       NULL|         3265.0|
| Valencia|    Deporte|         1550.0|


cubeCiudadCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, categoria: string ... 1 more field]

Pregunta 11 — Tabla dinámica con pivot

Crea una tabla dinámica donde las filas sean ciudades y las columnas sean categorías.

In [111]:
val pivotCiudadCategoria = marketNova
  .groupBy("ciudad")
  .pivot("categoria")
  .agg(sum("total"))
  .orderBy("ciudad")

pivotCiudadCategoria.show()

+---------+-------+-----------+------+
|   ciudad|Deporte|Electrónica| Hogar|
+---------+-------+-----------+------+
|Barcelona|  570.0|     3050.0| 730.0|
|   Bilbao|  850.0|     2100.0| 580.0|
|   Madrid| 1250.0|     4420.0| 900.0|
|  Sevilla| 1010.0|     1540.0|1000.0|
| Valencia| 1550.0|     1085.0| 630.0|
+---------+-------+-----------+------+



pivotCiudadCategoria: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [ciudad: string, Deporte: double ... 2 more fields]

Pregunta 12 — Guardar resultados en CSV

Guarda el resumen por ciudad en formato CSV.

In [112]:
pivotCiudadCategoria.write
  .mode("overwrite")
  .option("header", "true")
  .csv("C:/Curso-Scala/salida/resumen_ciudad")

Pregunta 13 —
Crea un pipeline que:

1. Calcule el resumen por categoria
2. Incluya:
    * número de ventas
    * unidades vendidas
    * facturación total
    * ticket medio
3. Ordene por facturación total de mayor a menor
4. Guarde el resultado en CSV en la ruta:

In [114]:
val resultado = marketNova
  .groupBy("categoria")
  .agg(
    count("id").as("num_ventas"),
    sum("total").as("total_facturado"),
    avg("total").as("ticket_medio"),
    max("total").as("venta_maxima")
  )
  .orderBy($"total_facturado".desc)

resultado.write
  .mode("overwrite")
  .option("header", "true")
  .csv("C:/Curso-Scala/salida/marketnova_resumen_categoria")

resultado: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [categoria: string, num_ventas: bigint ... 3 more fields]